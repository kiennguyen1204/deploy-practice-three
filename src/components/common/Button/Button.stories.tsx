import type { Meta, StoryObj } from '@storybook/react';

// constants
import { BUTTON_VARIANT } from 'constants/enums';

// components
import Button from 'components/common/Button';

const meta: Meta<typeof Button> = {
  component: Button
};

type Story = StoryObj<typeof Button>;

export const Primary: Story = {
  args: {
    label: 'Button',
    variant: BUTTON_VARIANT.PRIMARY
  }
};

export const Secondary: Story = {
  render: () => <Button label='Button' variant={BUTTON_VARIANT.SECONDARY} onClick={() => {}} />
};

export default meta;
