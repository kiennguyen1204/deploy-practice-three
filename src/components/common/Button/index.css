.btn-primary {
  display: inline-flex;
  height: var(--height-huge);
  justify-content: center;
  align-items: center;
  gap: 8px;
  flex-shrink: 0;
  border-radius: var(--border-radius-huge);
  background: var(--color-primary-base);
  border: none;
}

.btn-secondary {
  display: flex;
  height: var(--height-large);
  padding: 0px 20px;
  justify-content: center;
  align-items: center;
  gap: 12px;
  border-radius: var(--border-radius-huge);
  border: 1px solid var(--color-secondary-dark);
  background: var(--color-secondary-ultra-light);
}

.btn-color-text-primary {
  color: var(--color-neutral-white);
  font: normal var(--font-weight-regular) var(--font-size-normal) / var(--line-height-medium);
}

.btn-color-text-secondary {
  color: var(--color-neutral-mid);
  font: normal var(--font-weight-regular) var(--font-size-medium) / var(--line-height-medium);
}

.btn.btn-xs {
  padding: 0 20px;
}

.btn.btn-sm {
  padding: 0 24px;
}

.btn.btn-xl {
  padding: 0 40px;
}

.btn.btn-lg {
  padding: 8px 20px;
}

.btn.btn-xxl {
  padding: 16px 44px;
}

.btn.btn-big {
  padding: 8px 56px;
}

.btn.btn-bigger {
  padding: 24px 44px;
}
