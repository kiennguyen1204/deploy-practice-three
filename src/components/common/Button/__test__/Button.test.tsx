import { render, screen, fireEvent } from '@testing-library/react';

// constants
import {
  BUTTON_COLOR,
  BUTTON_COLOR_TEXT,
  BUTTON_SIZE,
  BUTTON_TYPE,
  BUTTON_VARIANT
} from 'constants/enums';

// components
import Button, { Props } from 'components/common/Button';

const setup = (props = {}) => {
  const defaultProps: Props = {
    label: 'Submit',
    variant: BUTTON_VARIANT.PRIMARY,
    onClick: jest.fn(),
    isLoading: false,
    ...props
  };

  render(<Button {...defaultProps} />);
};

describe('Button component', () => {
  test('renders button correctly', () => {
    setup();
    const buttonElement = screen.getByText('Submit');
    expect(buttonElement).toMatchSnapshot();
  });

  test('renders button with secondary variant', () => {
    setup({ variant: BUTTON_VARIANT.SECONDARY });

    const buttonElement = screen.getByText('Submit');
    expect(buttonElement).toMatchSnapshot();
  });

  test('renders button with primary variant', () => {
    setup({ variant: BUTTON_VARIANT.PRIMARY });

    const buttonElement = screen.getByText('Submit');
    expect(buttonElement).toMatchSnapshot();
  });

  it('renders with default variant (PRIMARY)', () => {
    setup({ variant: undefined });
    const buttonElement = screen.getByRole('button');
    expect(buttonElement).toHaveClass('btn-primary');
  });

  test('calls onClick handler when clicked', () => {
    const handleClick = jest.fn();
    setup({ onClick: handleClick });

    const buttonElement = screen.getByText('Submit');
    fireEvent.click(buttonElement);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test('renders button with loading spinner when isLoading is true', () => {
    setup({ isLoading: true });

    const loaderElement = screen.getByRole('button');
    expect(loaderElement).toBeInTheDocument();
  });

  it('should be disabled when isDisabled prop is true', () => {
    const onClick = jest.fn();
    const { getByRole } = render(
      <Button
        label='Test Button'
        isDisabled={true}
        onClick={onClick}
        variant={BUTTON_VARIANT.PRIMARY}
      />
    );
    const button = getByRole('button');
    fireEvent.click(button);
    expect(button).toBeDisabled();
    expect(onClick).not.toHaveBeenCalled();
  });

  it('should not be disabled when isDisabled prop is false', () => {
    const onClick = jest.fn();
    const { getByText } = render(
      <Button
        label='Test Button'
        isDisabled={false}
        onClick={onClick}
        variant={BUTTON_VARIANT.PRIMARY}
      />
    );
    const button = getByText('Test Button');
    fireEvent.click(button);
    expect(button).not.toBeDisabled();
    expect(onClick).toHaveBeenCalled();
  });

  it('renders a small button when size is SMALL', () => {
    setup({ size: BUTTON_SIZE.SMALL });
    const buttonElement = screen.getByRole('button');
    expect(buttonElement).toHaveClass('btn-sm');
  });

  it('renders a blue button when color is BLUE', () => {
    setup({ color: BUTTON_COLOR.SUCCESS });
    const buttonElement = screen.getByRole('button');
    expect(buttonElement).toHaveClass('btn-color-success');
  });

  it('renders a button with white text when colorText is WHITE', () => {
    setup({ colorText: BUTTON_COLOR_TEXT.PRIMARY });
    const buttonElement = screen.getByRole('button');
    expect(buttonElement).toHaveClass('btn-color-text-primary');
  });

  it('renders a submit button when type is SUBMIT', () => {
    setup({ type: BUTTON_TYPE.SUBMIT });
    const buttonElement = screen.getByRole('button');
    expect(buttonElement).toBeInTheDocument();
  });

  it('disables the button when isLoading is true', () => {
    const onClickMock = jest.fn();
    setup({ isLoading: true, onClick: onClickMock });
    const buttonElement = screen.getByRole('button');
    expect(buttonElement).toBeDisabled();
    fireEvent.click(buttonElement);
    expect(onClickMock).not.toHaveBeenCalled();
  });
});
