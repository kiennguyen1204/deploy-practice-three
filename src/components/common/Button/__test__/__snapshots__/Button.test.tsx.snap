// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders button correctly 1`] = `
<span>
  Submit
</span>
`;

exports[`Button component renders button with primary variant 1`] = `
<span>
  Submit
</span>
`;

exports[`Button component renders button with secondary variant 1`] = `
<span>
  Submit
</span>
`;
